encodings/protobuf/README.md

```diff
@@ -0,0 +1,225 @@
+types, the corresponding protobuf descriptor will mark them explicity
+as `optional`.  This has implications for most SDKs generated by
+`protoc`.  For example, in Go, the field of the struct becomes of
+pointer type, adding an extra indirection with the benefit that the
+presence or absence of the field can be consistently determined.
+
+Required attributes in OCSF are not marked as optional.  In most
+cases, an SDK will still permit distinguishing between present and
+absent values because the zero value is not a valid value anyway.  For
+example, `class_id` is required and non-zero; if the application finds
+it to have a zero value, the distinction between missing and invalid
+is not significant because missing _is_ invalid for required types.
+
+(Note that OCSF's JSON encoding does not distinguish between `null`
+and missing values, since null is not a type in OCSF)
+
+### Structured Types
+
+Structured types in OCSF include objects, events (or classes), and
+arrays.
+
+### Object and Event Types
+
+Objects and Events are mapped to Protobuf `message` objects in the
+straightforward way, preserving the OCSF name of the field (e.g.,
+`status_id` instead of the more conventional Protobuf `statusId`).
+This is so that the JSON encoding (with the option **Use proto field
+name instead of lowerCamelCase name** enabled) matches the OCSF JSON
+encoding.
+
+### The Un-extended Object
+
+The one exception to how objects are mapped, is the _unextended_
+`object`.  This is used in a few places in the schema (such as the
+`unmapped` attribute) to represent a schema-free JSON object.  For the
+Protobuf encoding, `object` is mapped to the well-known
+[Struct](https://protobuf.dev/reference/protobuf/google.protobuf/#struct)
+type.
+
+This has the desirable property that the native protobuf JSON encoding
+knows how to serialize and deserialize these types into plain JSON.
+
+Note that primitive JSON values such as string and number, and JSON
+arrays, are **NOT** valid values for an `object`.
+
+#### Field IDs
+
+The Protobuf wire encoding requires consistent field ids for message
+objects.  To facilitate interoperation and schema evolution, a
+separate control file (the canonical version of which is included in
+this repo in `control.json`) is used "remember" the field ids assigned
+to fields in messages.
+
+### Enumerated Values
+
+Enumerated **integer** values are built using protobuf's `enum`
+declaration.  Because OCSF often (not always) refines the set of
+enumerands on a per-event basis, the protobuf encoding supports that
+by making the enumeration types sub-names of the message types.
+Furthermore, because Protobuf uses "C++" scoping rules, in which
+enumerated values are in the same scope as the enum type itself
+(rather than children of the type), we mangle the enumerand names as
+well.
+
+For example:
+```
+message FileActivity {
+    enum ActionId {
+        UNKNOWN_ActionId = 0;  // The action was unknown. The <code>disposition_id</code>
```

Comment:
Maybe I'm weird, but I usually the form `<kind>_<field>` for cases like this, for example `ActivityId_UNKNOWN`. I just feel putting fixed part as a prefix makes constants like this stand out more and easier to mentally parse.

This is a stylistic concern, so _only_ make this changes if this matches your sense of taste (and, I suppose, norms in the Protobuf community).

Reply:
Ah; the example here is out of date with respect to the actual generated code, so worth updating!

The goal here is to follow idiomatic protobuf notation, which I think winds up being closer to your preference anyway.  For reference on the linting tool we're using, see https://buf.build/docs/lint/overview.